File: 9-palindrome-number/palindrome-number.cpp
class Solution {
public:
    bool isPalindrome(int x) {
        // Negative numbers aren't palindromes; numbers ending in 0 (except 0 itself) aren't either
        if (x < 0 || (x % 10 == 0 && x != 0)) {
            return false;
        }

        int reverted = 0;
        while (x > reverted) {
            reverted = reverted * 10 + x % 10;
            x /= 10;
        }

        // Even digit count: x == reverted
        // Odd digit count: middle digit doesn't matter, so drop it via reverted/10
        return x == reverted || x == reverted / 10;
    }
};

File: 9-palindrome-number/palindrome-number.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        # Negative numbers aren't palindromes; numbers ending in 0 (except 0 itself) aren't either
        if x < 0 or (x % 10 == 0 and x != 0):
            return False

        reverted = 0
        while x > reverted:
            reverted = reverted * 10 + x % 10
            x //= 10

        # Even digit count: x == reverted
        # Odd digit count: middle digit doesn't matter, so drop it via reverted // 10
        return x == reverted or x == reverted // 10
        